codec/src/main/java/io/netty/handler/codec/compression/AbstractClosableCompressionEncoder.java

```diff
@@ -0,0 +1,170 @@
+    }
+
+    @Override
+    public final long lastWriteTimeoutMillis() {
+        return lastWriteTimeoutUnit.toMillis(lastWriteTimeout);
+    }
+
+    /**
+     * {@inheritDoc}
+     * <p/>
+     * The default value is {@value #DEFAULT_LAST_WRITE_TIMEOUT}, unit is {@value #DEFAULT_LAST_WRITE_TIMEOUT_UNIT}.
+     */
+    @Override
+    public final void setLastWriteTimeout(long timeout, TimeUnit unit) {
+        lastWriteTimeout = timeout;
+        lastWriteTimeoutUnit = unit;
+    }
+
+    @Override
+    public abstract boolean isClosed();
+
+    @Override
+    public final ChannelFuture close() {
+        return close(ctx().newPromise());
+    }
+
+    @Override
+    public final ChannelFuture close(final ChannelPromise promise) {
+        final ChannelHandlerContext ctx = ctx();
+        EventExecutor executor = ctx.executor();
+        if (executor.inEventLoop()) {
+            return finishEncode(ctx, promise);
+        } else {
+            final ChannelPromise p = ctx.newPromise();
+            executor.execute(new OneTimeTask() {
+                @Override
+                public void run() {
+                    ChannelFuture f = finishEncode(ctx(), p);
+                    f.addListener(new ChannelPromiseNotifier(promise));
+                }
+            });
+            return p;
+        }
+    }
+
+    @Override
+    public final void close(final ChannelHandlerContext ctx, final ChannelPromise promise) throws Exception {
+        ChannelFuture f = finishEncode(ctx, ctx.newPromise());
+        if (f.isDone()) {
+            ctx.close(promise);
+        } else {
+            f.addListener(new ChannelFutureListener() {
+                @Override
+                public void operationComplete(ChannelFuture f) throws Exception {
+                    ctx.close(promise);
+                }
+            });
+            // Ensure the channel is closed even if the write operation completes in time
+            ctx.executor().schedule(new OneTimeTask() {
+                @Override
+                public void run() {
+                    if (!promise.isDone()) {
+                        ctx.close(promise);
+                    }
+                }
+            }, lastWriteTimeout, lastWriteTimeoutUnit);
+        }
+    }
+
+    /**
+     * Finishes current encoding.
+     *
+     * Note: for some {@link CompressionFormat}'s which does not support finish operation
+     * this method could fail specified {@link ChannelPromise}.
+     */
+    protected abstract ChannelFuture finishEncode(ChannelHandlerContext ctx, ChannelPromise promise);
+
+    /**
+     * Holds link to current {@link ChannelHandlerContext}.
+     */
+    @Override
+    public void handlerAdded(ChannelHandlerContext ctx) throws Exception {
+        this.ctx = ctx;
+    }
+
+    /**
+     * Discards link to current {@link ChannelHandlerContext}.
+     */
+    @Override
+    public void handlerRemoved(ChannelHandlerContext ctx) throws Exception {
+        this.ctx = null;
```

Comment:
Am I right that we also have to discard link to current context if handler removes from pipeline?